packages/api/src/main.ts

```diff
@@ -3,51 +3,57 @@
  * This is only a minimal backend to get started.
  */
 import * as express from 'express';
-import { openfeature } from '@openfeature/openfeature-js';
+import {
+  FlagEvaluationDetails,
+  openfeature,
+} from '@openfeature/openfeature-js';
 import { fibonacci } from '@openfeature/fibonacci';
 import { query, validationResult } from 'express-validator';
 import { OpenTelemetryHook, LoggingHook } from '@openfeature/extra';
 import { buildMarkup } from './markup';
 
 const app = express();
-const appName = 'api'
+const appName = 'api';
 
 openfeature.registerHooks(new OpenTelemetryHook(appName));
 const client = openfeature.getClient(appName);
 client.registerHooks(new LoggingHook());
 
 app.get('/api', async (req, res) => {
-  const message = (await client.isEnabled(
-    'new-welcome-message',
-    false
-  ))
+  const message = (await client.isEnabled('new-welcome-message', false))
     ? 'Welcome to the next gen api!'
     : 'Welcome to the api!';
   res.send({ message });
 });
 
 app.get('/hello', async (req, res) => {
-  const hexColor = (await client.getStringValue(
+  const hexColor = await client.getStringValue(
     'hex-color',
     '000000',
     undefined,
     {
       hooks: [
         {
-          after: (hookContext, flagValue: string) => {
+          name: '',
+          after: (
+            hookContext,
+            evaluationDetails: FlagEvaluationDetails<string>
+          ) => {
             // validate the hex value.
             const hexPattern = /[0-9A-Fa-f]{6}/g;
-            if (hexPattern.test(flagValue)) {
-              return flagValue;
+            if (hexPattern.test(evaluationDetails.value)) {
```

Comment:
One of the things I've been trying to bat around in my head is whether we _always_ want to have people use this wrapper.

Sometimes, I imagine you just want the value and you don't care why. That's probably even the most common case. Is it worth having the additional API surface value? I lean towards yes, but I'm not sure I've entirely convinced myself either.

Reply:
I agree with this, many end consumers of the flag evaluaitons only want a the value.

Reply:
I think I agree. I guess the question is how to best accomplish what you describe. I guess the most simple solution would be to change the `after` hook signature to take  the flagValue as well as the Evaluation details, so: `(hookContext, flagValue: T, evaluationDetails: EvaluationDetails<T>)`

Obviously, the implementer need not touch the evaluation details.

Reply:
@toddbaert Yes. I'm thinking I want these two mechanisms. Most folks will probably use option 1 90% of the time. I think I'm even fine if we implement option 1 via option 2.

```js
if (await client.getBooleanValue('show-redesign', false, ctx) == true) {
  showRedesign()
} else {
  showOldDesign();
}
```


```js
var eval = await client.getBooleanEvaluation('show-redesign', false, ctx);
if (eval.value == true) {
  log.info('selected because of: ', eval.reason);
  showRedesign()
} else {
  showOldDesign();
}
```

Reply:
@justinabrahms That's exactly what's in this PR :grin:  . These methods have been added to the "application-author"-facing API, all the "value-oriented" ones still exist. This might be helpful: https://github.com/open-feature/sdk-research/blob/8fb9e1ce9bba9b0afdd3df6ff06aa2b751fa72f1/README.md

Reply:
Because I think it's useful to think through, @s-sen and I were recently talking about what this would all look like in a java world.

```java

public class Example {

  public class MyHook implements Hook<T> {
    public void before(HookContext ctx) {
      // ...
    }
    public void after(HookContext ctx, FlagEvaluationDetails<T> details) {
      // ...
    }
  }

  public void myExample() {
    val ctx = Context.Builder()
      .addString("name", "R. U. There")
      .build();

    assert client.isEnabled("my-boolean-flag", false);
    String value = client.getStringValue("my-string-key", "hello, world!", ctx);

    EvaluationDetails<String> details = client.getStringDetails("my-string-key", "hello, world", ctx);
    String value2 = details.value;
    log.info("result was %s because %s", value2, details.reason);

    // example with hooks
    EvaluationDetails<String> hookDetails = client.getStringDetails("my-string-key", "hello, world", ctx, new MyHook());
  }
}
```

This opens up a few questions for me:
1. If a hook is registered globally, how will it know what the incoming data type is and if it can handle it?
2. I believe there are complexities for ensuring the java generics are the type that we think they'll be due to runtime type erasure.

Hopefully Shantanu will comment w/ more thoughts here.

Reply:
Great questions. In TS, we can supply a generic arg, but it's basically an assertion and might not be correct at runtime. In Java, one possible solution would be to use something like `Object` for a generic/global hook, which would work for at least all the boxed primitives.

A better solution might be to create a FlagValue reference type that can box any of the primitives.

Reply:
Here are my thoughts as I put on a developer hat

1. As I use the API, how quickly can the framework tell me if I am making a mistake - e.g. if I am using a statically typed language like Java, can the IDE tell me if I am adding an incorrect type for my default value?

This has been addressed by using different flavors of the API based on the type e.g. getValueAsBoolean

2. How can I easily troubleshoot simple issues if I am not getting the expected result e.g. misconfigured rule. Can I figure this out from any simple evaluation reason e.g. the LD enum and an accompanying brief string that indicates what the issue might be? This is my first step of troubleshooting before going deep into provider specific logs.

If we want to solve the above by using callback hooks, then I have to implement the hook interface and the put the implementation code somewhere that is feasible for me to access if I want to step through the hook callbacks in a debugger.  This seems more work on my part.

If we want to solve this by a 2nd flavor of the API that returns diagnostics information in addition to the value, then I would always use this flavor of the API - why will I use the flavor of the API that returns just the value and not diagnostics info?

Hence instead of a hook for diagnostics. Can we use a Wrapper object that wraps the value and the diagnostics so that the single API can return both and my code is more compact - I extract the value and log simple diagnostics in the same flow rather than use a hook and the associated callback code.

Thoughts?

Reply:
>Can we use a Wrapper object that wraps the value and the diagnostics so that the single API can return both and my code is more compact - I extract the value and log simple diagnostics in the same flow rather than use a hook and the associated callback code.

Agreed. I don't think the main purpose of hooks is for diagnostics. This PR actually adds exactly the type of methods you describe, which wrap the flag value with additional diagnostic information (for example, in addition to `getBooleanValue` it now adds `getBooleanDetails` which returns an `EvaluationDetails<boolean>` which returns the data you describe.

Hooks are for a different purpose than this, they are for adding additional behavior to flag evaluation, be it telemetry, validation, etc, which is something we've heard is critically important for potential users and integrators.

Does that address your concern?

Reply:
@s-sen @justinabrahms Can I consider this thread resolved? I really appreciate the feedback, and there's a need to keep things moving. If my comments haven't satisfied your concerns, please let me know!

Reply:
I feel pretty satisfied. The hooks are a powerful concept. I want to be able to also pass them in on the client level (not only the invocation level), but I think that's also covered.

Reply:
@justinabrahms Yep! currently, hooks can be registered at flag evaluation (most likely would be an anonymous class or a lambda in that case, but not necessarily), at the client level, or globally at the API level.

client: https://github.com/open-feature/sdk-research/blob/07c088d2568f956ec32b947aa7b47f83c7761dfe/packages/openfeature-js/src/lib/client.ts#L41

api: https://github.com/open-feature/sdk-research/blob/07c088d2568f956ec32b947aa7b47f83c7761dfe/packages/openfeature-js/src/lib/api.ts#L42

There's still a few open questions here, but there's some interesting use-cases for instantiating multiple clients, or even having request-scoped clients, so having hooks at both these levels seems like a good idea. I'm interested in some more conversations around this, I'm sure they'll come up as we start to define the spec.

Reply:
@toddbaert  So, it seems we are converging on is a spec with 2 flavors of the API - one that returns the value and the other the value and the diagnostics details. I am fine with this approach.